In [23]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import GridSearchCV, train_test_split, cross_val_score
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from plotnine import *

# Cleaning NBA and NCAA data

In [24]:
df_draft_raw = pd.read_csv("../data/draft.csv", index_col=0)

In [25]:
df_draft = (df_draft_raw.rename(columns={'Lg': 'league', 'Rd': 'round', 
                                         'Pk': 'pick', 'Tm': 'team', 'Pos': 'position'})
              .assign(Player = lambda x: x.Player.str.split('\\').str.get(0)))
df_draft.columns = df_draft.columns.str.lower()

In [26]:
df_draft.head()

year league  round  pick team          player     age position born  \
Rk                                                                        
1   2016    NBA      1     1  PHI     Ben Simmons  19.343      G-F   au   
2   2016    NBA      1     2  LAL  Brandon Ingram  18.299        F   us   
3   2016    NBA      1     3  BOS    Jaylen Brown  19.247      F-G   us   
4   2016    NBA      1     4  PHO   Dragan Bender  18.223        F   ba   
5   2016    NBA      1     5  MIN       Kris Dunn  22.103        G   us   

                       college  
Rk                              
1   Louisiana State University  
2              Duke University  
3     University of California  
4                          NaN  
5           Providence College

In [27]:
print("There are {} years, {} teams, {} players".format(len(pd.unique(df_draft.year)),
                                                        len(pd.unique(df_draft.team)), 
                                                        len(pd.unique(df_draft.player))))

There are 19 years, 37 teams, 1122 players


There are players who are not drafted form any college. They mostly are non-US players.

In [28]:
df_draft.loc[pd.isnull(df_draft.college), :]

year league  round  pick team                player     age position  \
Rk                                                                           
4     2016    NBA      1     4  PHO         Dragan Bender  18.223        F   
10    2016    NBA      1    10  MIL            Thon Maker  19.123      F-C   
13    2016    NBA      1    13  PHO  Georgios Papagiannis  18.360        C   
15    2016    NBA      1    15  DEN      Juan Hernangomez  20.274        F   
16    2016    NBA      1    16  BOS    Guerschon Yabusele  20.194        F   
23    2016    NBA      1    23  BOS            Ante Zizic  19.175      F-C   
24    2016    NBA      1    24  PHI        Timothe Luwawu  21.051        F   
26    2016    NBA      1    26  PHI        Furkan Korkmaz  18.339        G   
32    2016    NBA      2    32  LAL           Ivica Zubac  19.102        C   
35    2016    NBA      2    35  BOS          Rade Zagorac  20.321      G-F   
39    2016    NBA      2    39  NOP       David Michineau  22.023        G   
43    2016    NBA      2    43  HOU               Zhou Qi  20.164      F-C   
44    2016    NBA      2    44  ATL       Isaia Cordinier  19.212        G   
48    2016    NBA      2    48  CHI           Paul Zipser  22.131      G-F   
53    2016    NBA      2    53  DEN         Petr Cornelie  20.338        F   
57    2016    NBA      2    57  MEM           Wang Zhelin  22.160        C   
64    2015    NBA      1     4  NYK    Kristaps Porzingis  19.332      F-C   
65    2015    NBA      1     5  ORL         Mario Hezonja  20.125      F-G   
67    2015    NBA      1     7  DEN       Emmanuel Mudiay  19.116        G   
86    2015    NBA      1    26  SAS      Nikola Milutinov  20.182        C   
91    2015    NBA      2    31  MIN            Cedi Osman  20.083        F   
95    2015    NBA      2    35  PHI     Willy Hernangomez  21.034        C   
99    2015    NBA      2    39  CHO     Juan Pablo Vaulet  19.099        F   
107   2015    NBA      2    47  PHI      Arturas Gudaitis  22.011        C   
110   2015    NBA      2    50  ATL       Marcus Eriksson  21.207        G   
112   2015    NBA      2    52  DAL          Satnam Singh  19.202        C   
114   2015    NBA      2    54  UTA             Dani Diez  22.084        F   
117   2015    NBA      2    57  DEN      Nikola Radicevic  21.066        G   
119   2015    NBA      2    59  ATL   Dimitrios Agravanis  20.192      F-C   
120   2015    NBA      2    60  PHI         Luka Mitrovic  22.101        F   
...    ...    ...    ...   ...  ...                   ...     ...      ...   
931   2001    NBA      2    38  DET           Mehmet Okur  22.038      C-F   
940   2001    NBA      2    47  DEN         Ousmane Cisse     NaN      NaN   
941   2001    NBA      2    48  MEM        Antonis Fotsis  20.092        F   
949   2001    NBA      2    56  SAS     Robertas Javtokas  21.104        C   
954   2000    NBA      1     3  LAC          Darius Miles  18.267        F   
967   2000    NBA      1    16  SAC         Hedo Turkoglu  21.107        F   
974   2000    NBA      1    23  UTA     DeShawn Stevenson  19.091        G   
975   2000    NBA      1    24  CHI       Dalibor Bagaric  20.147        C   
976   2000    NBA      1    25  PHO       Jake Tsakalidis  21.024        C   
978   2000    NBA      1    27  IND         Primoz Brezec  20.275        C   
981   2000    NBA      2    30  LAC           Marko Jaric  21.265        G   
987   2000    NBA      2    36  NJN       Soumaila Samake  22.108        C   
993   2000    NBA      2    42  SEA       Olumide Oyedeji  19.053        C   
998   2000    NBA      2    47  SEA           Josip Sesar     NaN      NaN   
1002  2000    NBA      2    51  MIN        Igor Rakocevic  22.097        G   
1014  1999    NBA      1     5  TOR       Jonathan Bender  18.155        F   
1024  1999    NBA      1    15  NYK         Frederic Weis     NaN      NaN   
1033  1999    NBA      1    24  UTA      Andrei Kirilenko  18.136        F   
1038  1999    NBA      1    

Player names are unique, and ~60 are drafted each year. The number of colleges are different across years because some players are not drafted from college (e.g. from high school, from other countries.)

In [29]:
df_draft.groupby('year')['player', 'college'].agg(['count', 'nunique'])

player         college        
      count nunique   count nunique
year                               
1998     58      58      50      39
1999     58      58      51      45
2000     58      58      47      36
2001     57      57      45      35
2002     57      57      42      32
2003     58      58      33      31
2004     59      59      37      33
2005     60      60      37      32
2006     60      60      44      30
2007     60      60      47      35
2008     60      60      49      35
2009     60      60      48      35
2010     60      60      53      40
2011     60      60      46      34
2012     60      60      51      31
2013     60      60      47      40
2014     60      60      46      32
2015     60      60      46      32
2016     60      60      44      31

In [30]:
df_nba_raw = pd.read_csv("../data/nbarookiedata.csv", skiprows=1, index_col=0)

In [31]:
df_nba = (df_nba_raw
             .assign(Player = lambda x: x.Player.str.split('\\').str.get(0),
                     Season = lambda x: pd.to_numeric(x.Season.str.split('-').str.get(0)))
             [['Player', 'Tm', 'Season', 'PTS', 'TRB', 'AST', 'STL', 'BLK']]
             .rename(columns={'Player': 'player', 'Tm': 'team', 'Season': 'year', 
                              'PTS': 'point', 'TRB': 'rebound', 'AST': 'assist',
                              'STL': 'steal', 'BLK': 'block'}))

In [32]:
df_nba.head()

player team  year  point  rebound  assist  steal  block
Rk                                                                  
1      Alex Abrines  OKC  2016    6.0      1.3     0.6    0.5    0.1
2         Ron Baker  NYK  2016    4.1      1.9     2.1    0.7    0.2
3      Wade Baldwin  MEM  2016    3.2      1.4     1.8    0.5    0.2
4     Malik Beasley  DEN  2016    3.8      0.8     0.5    0.3    0.0
5   DeAndre' Bembry  ATL  2016    2.7      1.6     0.7    0.2    0.1

In [33]:
df_final = pd.merge(df_draft, df_nba, how='left', on=['player', 'team', 'year'])

In [22]:
# There are a lot of players drafted but don't appear in the NBA dataset, why?
df_final.loc[pd.isnull(df_final.point), :]

NameError: name 'df_final' is not defined

In [10]:
df_ncaa_raw = pd.read_csv('../data/ncaa2.csv', index_col=0)

In [11]:
df_ncaa_raw.columns

Index(['Name', 'Team', 'GP', 'Min', 'Pts', 'FG', 'FGA', 'FG%', '2Pt', '2PtA',
       '2P%', '3Pt', '3PtA', '3P%', 'FTM', 'FTA', 'FT%', 'Off', 'Def', 'TOT',
       'Asts', 'Stls', 'Blks', 'TOs', 'PFs', 'year', 'PTs/g', 'FGA/g',
       'Pts/Play', 'TS%', 'eFG%', 'FTA/FGA', '3PA/FGA', 'Ast/g', 'Ast/FGA',
       'A/TO', 'PPR', 'BK/g', 'STL/g', 'PF/g'],
      dtype='object')

In [12]:
df_ncaa = (df_ncaa_raw
              [['Name', 'Team', 'year', 'Pts', 'TOT', 'Asts', 'Stls', 'Blks']]
              .rename(columns={'Name': 'player', 'Team': 'college', 'year': 'col_year',
                               'Pts': 'col_point', 'TOT': 'col_rebound', 'Asts': 'col_assist',
                               'Stls': 'col_steal', 'Blks': 'col_block'}))

In [13]:
df_ncaa.head()

player           college  col_year  col_point  col_rebound  \
0       Luis Rivas  Western Illinois      2002       27.9         10.7   
1   Steve Reynolds  Western Michigan      2002       27.8          6.5   
2   Michael Watson              UMKC      2002       24.9          3.6   
3     Tyrone Hayes             Idaho      2002       34.3         12.3   
4  Travis Robinson      Jacksonville      2002       27.4          7.5   

   col_assist  col_steal  col_block  
0         0.0        0.0        0.5  
1         2.0        1.2        0.1  
2         3.2        2.1        0.2  
3         0.0        2.5        2.5  
4         1.1        1.9        0.3

In [14]:
df_final = pd.merge(df_final, df_ncaa, how='inner', on=['player'])

In [15]:
# Only keep the last year of college performance
df_final = (df_final.groupby('player')
        .apply(lambda df: df.sort_values(by='col_year', ascending=False).iloc[0, :]))

In [16]:
print("Number of unique players: {}".format(len(pd.unique(df_final.player))))

Number of unique players: 613


# Model building

In [17]:
df_final = df_final.dropna()
X = df_final[['round', 'pick', 'position',
              'col_point', 'col_rebound', 'col_assist', 'col_steal', 'col_block']]
X = pd.get_dummies(X, columns=['position'])
y = df_final['point']

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

## Random Forest Regression

In [19]:
rf = RandomForestRegressor(criterion='mse', n_jobs=-1)

## Tuning and choosing the best hyperparams

In [20]:
gs = GridSearchCV(estimator=rf, 
                  param_grid=[{'n_estimators': [100, 1000, 1000],
                               'min_samples_leaf': [1, 10, 100]}],
                  cv=10, scoring='neg_mean_squared_error')
gs.fit(X_train, y_train)

GridSearchCV(cv=10, error_score='raise',
       estimator=RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=-1,
           oob_score=False, random_state=None, verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid=[{'n_estimators': [100, 1000, 1000], 'min_samples_leaf': [1, 10, 100]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring='neg_mean_squared_error', verbose=0)

In [21]:
print(gs.best_params_)
best_rf = gs.best_estimator_
best_rf.fit(X_train, y_train)

{'min_samples_leaf': 10, 'n_estimators': 1000}


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=10, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=1000, n_jobs=-1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

## Model performance evaluation

In [26]:
print("Min / Median / Max of Point: {} {} {}".format(np.min(df_final.point),
                                                   np.median(df_final.point),
                                                   np.max(df_final.point)))
print("Train MSE: {:.3f}; Test MSE: {:.3f}".format(mean_squared_error(y_train, best_rf.predict(X_train)),
                                                 mean_squared_error(y_test, best_rf.predict(X_test))))

Min / Median / Max of Point: 0.0 4.6 21.0
Train MSE: 6.798; Test MSE: 9.499


The Train MSE is lower than Test MSE substantially, so it seems like our model overfits the training data. Also the error is substantial given that the median point scored is only 4.6.

We will build better model, potentially including team statistics along with individual statistics.